twake/backend/node/src/services/messages/services/messages/utils.ts

```diff
@@ -0,0 +1,83 @@
+import { getInstance, Message, MessageReaction } from "../../entities/messages";
+import { ThreadExecutionContext } from "../../types";
+
+export function getSubtype(
+  item: Pick<Message, "subtype">,
+  context?: ThreadExecutionContext,
+): null | "application" | "deleted" | "system" {
+  //Application request
+  if (context?.user?.application_id) {
+    return item.subtype === "application" ? "application" : null;
+  }
+  //System request
+  else if (context?.user?.server_request) {
+    return item.subtype;
+  }
+
+  //User cannot set a subtype itself
+  return null;
+}
+
+export function updateMessageReactions(
+  message: Message,
+  selectedReactions: string[],
+  userId: string,
+) {
+  let reactions: { [key: string]: MessageReaction } = {};
+  for (const reaction of message.reactions || []) {
+    reactions[reaction.name] = reaction;
+  }
+  for (const reaction of selectedReactions) {
+    reactions[reaction] = reactions[reaction] || { name: reaction, count: 0, users: [] };
+  }
+  for (const key in reactions) {
+    if (reactions[key].users.includes(userId)) {
+      reactions[key].count--;
+      reactions[key].users = reactions[key].users.filter(u => u != userId);
+    }
+    if (selectedReactions.includes(key)) {
+      reactions[key].count++;
+      reactions[key].users.push(userId);
+    }
+    if (reactions[key].count === 0) {
+      delete reactions[key];
+    }
+  }
+  message.reactions = Object.values(reactions);
+}
```

Comment:
Probably better ways of doing this (@chamerling is going to 🔪  me because of not using map)

And yeah, the guy comments on its own PR

Reply:
AHHHHHHHHHHHHH!!!!!!